platform/javascript/javascript_singleton.cpp

```diff
@@ -0,0 +1,338 @@
+	JavaScriptObjectImpl() {}
+	JavaScriptObjectImpl(int p_id) { _js_id = p_id; }
+	~JavaScriptObjectImpl() {
+		if (_js_id) {
+			godot_js_wrapper_object_unref(_js_id);
+		}
+	}
+};
+
+bool JavaScriptObjectImpl::_set(const StringName &p_name, const Variant &p_value) {
+	ERR_FAIL_COND_V_MSG(!_js_id, false, "Invalid JS instance");
+	const String name = p_name;
+	godot_js_wrapper_ex exchange;
+	void *lock = nullptr;
+	const Variant *v = &p_value;
+	int type = _variant2js((const void **)&v, 0, &exchange, &lock);
+	godot_js_wrapper_object_set(_js_id, name.utf8().get_data(), type, &exchange);
+	if (lock) {
+		_free_lock(&lock, type);
+	}
+	return true;
+}
+
+bool JavaScriptObjectImpl::_get(const StringName &p_name, Variant &r_ret) const {
+	ERR_FAIL_COND_V_MSG(!_js_id, false, "Invalid JS instance");
+	const String name = p_name;
+	godot_js_wrapper_ex exchange;
+	int type = godot_js_wrapper_object_get(_js_id, &exchange, name.utf8().get_data());
+	r_ret = _js2variant(type, &exchange);
+	return true;
+}
+
+Variant JavaScriptObjectImpl::getvar(const Variant &p_key, bool *r_valid) const {
+	if (r_valid) {
+		*r_valid = false;
+	}
+	godot_js_wrapper_ex exchange;
+	void *lock = nullptr;
+	const Variant *v = &p_key;
+	int prop_type = _variant2js((const void **)&v, 0, &exchange, &lock);
+	int type = godot_js_wrapper_object_getvar(_js_id, prop_type, &exchange);
+	if (lock) {
+		_free_lock(&lock, prop_type);
+	}
+	if (type < 0) {
+		return Variant();
+	}
+	if (r_valid) {
+		*r_valid = true;
+	}
+	return _js2variant(type, &exchange);
+}
+
+void JavaScriptObjectImpl::setvar(const Variant &p_key, const Variant &p_value, bool *r_valid) {
+	if (r_valid) {
+		*r_valid = false;
+	}
+	godot_js_wrapper_ex kex, vex;
+	void *klock = nullptr;
+	void *vlock = nullptr;
+	const Variant *kv = &p_key;
+	const Variant *vv = &p_value;
+	int ktype = _variant2js((const void **)&kv, 0, &kex, &klock);
+	int vtype = _variant2js((const void **)&vv, 0, &vex, &vlock);
+	int ret = godot_js_wrapper_object_setvar(_js_id, ktype, &kex, vtype, &vex);
+	if (klock) {
+		_free_lock(&klock, ktype);
+	}
+	if (vlock) {
+		_free_lock(&vlock, vtype);
+	}
+	if (ret == 0 && r_valid) {
+		*r_valid = true;
+	}
+}
+
+void JavaScriptObjectImpl::_get_property_list(List<PropertyInfo> *p_list) const {
+}
+
+void JavaScriptObjectImpl::_free_lock(void **p_lock, int p_type) {
+	ERR_FAIL_COND_MSG(*p_lock == nullptr, "No lock to free!");
+	const Variant::Type type = (Variant::Type)p_type;
+	switch (type) {
+		case Variant::STRING: {
+			CharString *cs = (CharString *)(*p_lock);
+			memdelete(cs);
+			*p_lock = nullptr;
+		} break;
+		default:
+			ERR_FAIL_MSG("Unknown lock type to free. Likely a bug.");
+	}
+}
+
+Variant JavaScriptObjectImpl::_js2variant(int p_type, godot_js_wrapper_ex *p_val) {
+	Variant::Type type = (Variant::Type)p_type;
+	switch (type) {
+		case Variant::BOOL:
+			return p_val->i ? Variant(true) : Variant(false);
```

Comment:
`Variant` is probably just as happy if you return `(bool)p_val->i` directly, but this is fine too :)

Reply:
I'll give it a try, as it would skip a an `if`.

Reply:
Yep, much better :)